compiler/AST/LoopExpr.cpp

```diff
@@ -204,6 +211,12 @@ bool LowerLoopExprVisitor::enterLoopExpr(LoopExpr* node) {
     // The iterator expr might be a loop-expr itself, make sure it gets
     // lowered.
     iterExpr->accept(this);
+
+    // Do not preserve the shape if there is a filtering predicate.
+    if (node->forall && noFilter) {
+      normalize(replacement); // for addIterRecShape()
```

Comment:
it seems preferable to me for addIterRecShape to deal with whatever form of AST it needs to handle and generate right here instead of adding the new normalize call.

Reply:
(scratch that, I was missing that LowerLoopExprVisitor runs during normalize)